model/model.js

```diff
@@ -1,8 +1,202 @@
 const db = require("../db/connection.js");
 
-exports.selectTopics = () => {
+const selectTopics = () => {
     return db.query("SELECT * FROM topics;").then((selectTopicsQueryResult) => {
         const selectTopicsResponseObject = {topics: selectTopicsQueryResult.rows};
         return selectTopicsResponseObject;
     });
-};
\ No newline at end of file
+};
+
+const selectArticles = (urlParams) => {
+    const lowercaseParams = {};
+    Object.keys(urlParams).forEach((element, index) => {
+        // default to lowercase
+        lowercaseParams[element.toLowerCase()] = urlParams[element].toLowerCase();
+    });
+
+
+    let isError = false;
+    let isErrorString = "";
+    let sortBy = "created_at";
+    let topicBool = false;
+    let topic = "";
+    let order = "desc"
+
+
+    if (urlParams) {
+                
+        if (lowercaseParams["sort_by"]) {
+            // check if a valid column was specified
+            const columnsList = ["title","topic","author","body","created_at","votes"];
+            if (columnsList.includes(lowercaseParams["sort_by"])) {
+                sortBy = lowercaseParams["sort_by"];
+            } else {
+                isError = true;
+                isErrorString = "invalid parameter was specified in sort_by";
+            };
+        };
+
+        if (lowercaseParams["order"]) {
+
+
+            const orderListAsc = ["asc",  "ascending",  "a"];
+            const orderListDesc = ["desc", "descending", "d"];
+
+            if (orderListAsc.includes(lowercaseParams["order"])) {
+                order = "asc";
+            } else if (orderListDesc.includes(lowercaseParams["order"])) {
+                order = "desc";
+            } else {
+                isError = true;
+                isErrorString = "";
+            };
+        };
+
+        if (lowercaseParams["topic"]) {
+            topicBool = true;
+            topic =   lowercaseParams["topic"];          
+
+        };
+
+    };       
+       
+        const sqlQuery = `SELECT articles.article_id, articles.title, articles.topic, articles.author, articles.created_at, articles.votes, CAST(COUNT(comments.*) AS int) AS comment_count FROM articles LEFT JOIN comments ON comments.article_id = articles.article_id GROUP BY articles.article_id ORDER BY ${sortBy} ${order};`;
+          
+    
+ 
+    return db.query(sqlQuery).then((queryResult) => {
+        
+        const responseObject = {};
+
+        // this is probably the wrong way to do this.. will remove once i figure out how to do it in an sql query. tried using WHERE in various different ways but none seemed to work with the JOIN 
+        if (topicBool) {
+            const topicFilteredResult = queryResult.rows.filter((element) => {
```

Comment:
Here you are filtering based on topic, this is a slow process in terms of processing, it would be better handled using SQL, to do this if you wanted I would recommend looking at the complex Queries notes